libgnucash/engine/ScrubBudget.c

```diff
@@ -0,0 +1,256 @@
+
+    PWARN ("heuristics_on_budget %s: A(%d) L(%d) Inc(%d) Exp(%d) Eq(%d) = %d",
+           gnc_budget_get_name (budget),
+           heuristics.asset, heuristics.liability, heuristics.income,
+           heuristics.expense, heuristics.equity, result);
+
+    return result;
+}
+
+static void fix_budget_acc_sign (Account *acc, gpointer user_data)
+{
+    guint i;
+    gnc_numeric amt;
+    Reversal_t* reversal = (Reversal_t*) user_data;
+    GncBudget* budget = reversal->budget;
+    guint numperiods = gnc_budget_get_num_periods (budget);
+    int type = xaccAccountGetType (acc);
+
+    switch (reversal->policy)
+        {
+        case 5:                 /* inc-exp */
+            if (!((type == ACCT_TYPE_INCOME) || (type == ACCT_TYPE_EXPENSE)))
+                return;
+            PWARN ("budget account [%s] is inc/exp. reverse!",
+                   xaccAccountGetName(acc));
+            break;
+        case 7:                 /* credit-accounts */
+            if (!((type == ACCT_TYPE_LIABILITY) || (type == ACCT_TYPE_PAYABLE) ||
+                  (type == ACCT_TYPE_EQUITY) || (type == ACCT_TYPE_INCOME) ||
+                  (type == ACCT_TYPE_CREDIT)))
+                return;
+            PWARN ("budget account [%s] is credit-account. reverse!",
+                   xaccAccountGetName(acc));
+            break;
+        }
+
+    for (i=0; i < numperiods; ++i)
+    {
+        if (!(gnc_budget_is_account_period_value_set (budget, acc, i)))
+            continue;
+
+        amt = gnc_budget_get_account_period_value (budget, acc, i);
+        amt = gnc_numeric_neg (amt);
+        gnc_budget_set_account_period_value (budget, acc, i, amt);
+        PWARN ("budget account [%s] period %d sign-reversed!",
+               xaccAccountGetName(acc), i + 1);
+    }
+
+    PWARN ("budget account reversal [%s] completed!", xaccAccountGetName(acc));
+}
+
+static void fix_budget_sign (QofInstance* data, gpointer user_data)
+{
+    GncBudget* budget = GNC_BUDGET(data);
+    Account *root = (Account*) user_data;
+    Reversal_t reversal;
+
+    /* output: 5=inc/exp, 6=none, 7=credit-accounts */
+    reversal.policy = heuristics_on_budget (budget, root);
+    reversal.budget = budget;
+
+    if (reversal.policy == 6)
+        PWARN ("processing budget [%s]: skipped because reversal=none",
+               gnc_budget_get_name (budget));
+    else
+    {
+        PWARN ("processing budget [%s] for reversal", gnc_budget_get_name (budget));
+        gnc_account_foreach_descendant (root, fix_budget_acc_sign, &reversal);
+    }
+}
+
+/* analyse book, and return following:
+ * 0: don't offer scrubbing
+ * 1: set the flag without scrubbing
+ * 2: offer scrubbing */
+guint scrub_budget_signs_check (QofBook *book)
+{
+    /* feature set. skip scrubbing. */
+    if (gnc_features_check_used (book, GNC_FEATURE_BUDGET_UNREVERSED))
+        return 0;
+
+    /* there are no budgets. set feature. */
+    if (!(gnc_budget_get_default (book)))
+        return 1;
```

Comment:
> Instead defer this to when the user makes changes to some budget value. In fact I think that's the only place where you should set it. Even scrub doesn't have to set it explicitly as it will also make changes to some budget values which then in turn will set the feature code for you.

This is not a bad idea. So leave book unchanged until budget-editor processes a budget-editor-cell input. Will need to think about that some more.